### PR TITLE
[pack] ANCM 2.1 integration

### PR DESCRIPTION
**Pull request by alrod**

### CODE REVIEW

File: build.ps1 [hunk 1]
```diff
@@ -50,26 +59,36 @@ function CrossGen([string] $runtime, [bool] $isSelfContained, [string] $publishT
...
 
-        & $crossGen $prm >> $buildOutput\crossgenout.$runtime.txt
+        & $crossGen $prm >> $buildOutput\crossgenout.$runtime.txt  2>&1
 
         $niDll = Join-Path $privateSiteExtensionPath $([io.path]::GetFileNameWithoutExtension($_.FullName) + ".ni.dll")
         if ([System.IO.File]::Exists($niDll)) {
             Remove-Item $_.FullName
             Rename-Item -Path $niDll -NewName $_.FullName
+
+            #& $crossGen "/Platform_Assemblies_Paths", "$selfContained", "/CreatePDB", "$symbolsPath", $_.FullName >> $buildOutput\crossgenout.$runtime.txt 2>&1
```
Comment (safihamid) [inline build.ps1]: Why is this commented out? wedo really need the pdbs

Comment (fabiocav) [reply build.ps1]: Either way, let's remove commented out code. Either by re-adding or deleting.

File: build.ps1 [hunk 2]
```diff
@@ -28,20 +28,29 @@ function CrossGen([string] $runtime, [bool] $isSelfContained, [string] $publishT
...
     $crossGen = "$publishTarget\download\crossgen\crossgen.exe"
 
-    #DownloadNupkg "https://dotnet.myget.org/F/dotnet-core/api/v2/package/runtime.$runtime.microsoft.netcore.jit/2.1.0-preview2-26316-09" @("runtimes\$runtime\native\clrjit.dll")  @("$publishTarget\download\clrjit")
-    #DownloadNupkg "https://dotnet.myget.org/F/dotnet-core/api/v2/package/runtime.$runtime.Microsoft.NETCore.Runtime.CoreCLR/2.1.0-preview2-26316-09"  @("tools\crossgen.exe")  @("$publishTarget\download\crossgen")
+	#$symbolsPath = Join-Path $publishTarget "Symbols"
+    #new-item -itemtype directory -path $symbolsPath
 
-    DownloadNupkg "https://www.nuget.org/api/v2/package/runtime.$runtime.Microsoft.NETCore.Jit/2.0.5" @("runtimes\$runtime\native\clrjit.dll")  @("$publishTarget\download\clrjit")
-    DownloadNupkg "https://www.nuget.org/api/v2/package/runtime.$runtime.Microsoft.NETCore.Runtime.CoreCLR/2.0.5"  @("tools\crossgen.exe")  @("$publishTarget\download\crossgen")
+    DownloadNupkg "https://dotnet.myget.org/F/dotnet-core/api/v2/package/runtime.$runtime.Microsoft.NETCore.Jit/2.1.0-rtm-26528-02" @("runtimes\$runtime\native\clrjit.dll")  @("$publishTarget\download\clrjit")
```
Comment (fabiocav) [inline build.ps1]: 👍 
Thought I had updated this with the change to 2.1. Likely a commit that was missed when the PR was broken up.

File: build.ps1 [hunk 3]
```diff
@@ -28,20 +28,29 @@ function CrossGen([string] $runtime, [bool] $isSelfContained, [string] $publishT
...
+    DownloadNupkg "https://dotnet.myget.org/F/dotnet-core/api/v2/package/runtime.$runtime.Microsoft.NETCore.Runtime.CoreCLR/2.1.0-rtm-26528-02"  @("tools\crossgen.exe")  @("$publishTarget\download\crossgen")
     DownloadNupkg "https://www.nuget.org/api/v2/package/Microsoft.Build.Tasks.Core/15.1.1012" @("lib\netstandard1.3\Microsoft.Build.Tasks.Core.dll")  @("$selfContained")
     DownloadNupkg "https://www.nuget.org/api/v2/package/Microsoft.Build.Utilities.Core/15.1.1012" @("lib\netstandard1.3\Microsoft.Build.Utilities.Core.dll")  @("$selfContained")
+    if ($runtime -eq "win-x86") {
+        DownloadNupkg "https://dotnet.myget.org/F/aspnetcore-dev/api/v2/package/Microsoft.AspNetCore.AspNetCoreModuleV2/2.1.0-a-oob-2-1-oob-16914" @("contentFiles\any\any\x86\aspnetcoreV2.dll", "contentFiles\any\any\x86\aspnetcoreV2_inprocess.dll") @("$privateSiteExtensionPath\ancm", "$privateSiteExtensionPath\ancm")
```
Comment (fabiocav) [inline build.ps1]: Wouldn't adding this dependency handle the content file copy to the application output? That would remove the need to explicitly download and extract those files as part of the build.

File: build.ps1 [hunk 4]
```diff
@@ -28,20 +28,29 @@ function CrossGen([string] $runtime, [bool] $isSelfContained, [string] $publishT
...
+    } else {
+        DownloadNupkg "https://dotnet.myget.org/F/aspnetcore-dev/api/v2/package/Microsoft.AspNetCore.AspNetCoreModuleV2/2.1.0-a-oob-2-1-oob-16914" @("contentFiles\any\any\x64\aspnetcoreV2.dll", "contentFiles\any\any\x64\aspnetcoreV2_inprocess.dll") @("$privateSiteExtensionPath\ancm", "$privateSiteExtensionPath\ancm")
+    }
+    DownloadNupkg "https://www.nuget.org/api/v2/package/Microsoft.Web.Xdt/2.1.2" @("lib\net40\Microsoft.Web.XmlTransform.dll")  @("$publishTarget\download\xdt")
```
Comment (fabiocav) [inline build.ps1]: This download/extraction can also be removed and we can simplify the transformation (by removing the custom implementation) if we install the `dotnet-xdt` global tool:
`dotnet tool install -g dotnet-xdt --version 2.1.0-rc.1`

File: build.ps1 [hunk 5]
```diff
@@ -28,20 +28,29 @@ function CrossGen([string] $runtime, [bool] $isSelfContained, [string] $publishT
...
     # Publish self-contained app with all required dlls for crossgen
     dotnet publish .\src\WebJobs.Script.WebHost\WebJobs.Script.WebHost.csproj -r $runtime -o "$selfContained" -v q /p:BuildNumber=$buildNumber    
 
     # For self contained we want to process only IL ddls
     if ($isSelfContained) {
-        dotnet publish .\src\WebJobs.Script.WebHost\WebJobs.Script.WebHost.csproj -o $regular -v q /p:BuildNumber=$buildNumber
+        dotnet publish .\src\WebJobs.Script.WebHost\WebJobs.Script.WebHost.csproj -o $regular -v q /p:BuildNumber=$buildNumber -c Release
+
+        # Modify web.config for inproc
+        XmlDocTransform "$privateSiteExtensionPath\web.config" "$privateSiteExtensionPath\web.InProcess.xdt" "$publishTarget\download\xdt\Microsoft.Web.XmlTransform.dll"
```
Comment (fabiocav) [inline build.ps1]: Once the xdt global tool is installed, you can just call `dotnet-xdt` with the parameters.

File: build.ps1 [hunk 6]
```diff
@@ -50,26 +59,36 @@ function CrossGen([string] $runtime, [bool] $isSelfContained, [string] $publishT
     Foreach-Object {
 
         # For self contained we want to process only IL ddls
+        <#
```
Comment (fabiocav) [inline build.ps1]: Remove commented out code if not needed.

File: build.ps1 (hunk 3 above)
Comment (safihamid) [inline build.ps1]: >org [](start = 44, length = 3)

This wouldn't work for x64 unless we assume V2 will be broken for x64 and address this in a separate PR.

File: build.ps1 [hunk 7]
```diff
@@ -116,6 +138,32 @@ function DownloadNupkg([string] $nupkgPath, [string[]]$from, [string[]]$to) {
     }
 }
 
+function XmlDocTransform($xml, $xdt, $xdtDll)
```
Comment (fabiocav) [inline build.ps1]: If the xdt tool mentioned above is used, this should no longer be needed.

File: build.ps1 [hunk 8]
```diff
@@ -185,5 +233,5 @@ if ($bypassPackaging){
     BuildPackages "" $false
 
     #build win-x86 extensions
-    BuildPackages "win-x86" $false
+    BuildPackages "win-x86" $true
```
Comment (fabiocav) [inline build.ps1]: So this will effectively change the bits we're running in prod to no longer use the shared runtime, correct?

File: src/WebJobs.Script.WebHost/Program.cs [hunk 9]
```diff
@@ -22,7 +22,7 @@ public static void Main(string[] args)
 
         public static IWebHost BuildWebHost(string[] args)
         {
-            return CreateWebHostBuilder(args).Build();
+            return CreateWebHostBuilder(args).UseIIS().Build();
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Program.cs]: This should be conditional on the environment. When running locally, we don't need this set.

File: src/WebJobs.Script.WebHost/Resources/Functions/WarmUp/run.csx [hunk 10]
```diff
@@ -1,13 +1,9 @@
-﻿using System.Net;
+﻿using Microsoft.AspNetCore.Mvc;
 
-public static Task<HttpResponseMessage> Run(HttpRequestMessage req, TraceWriter log)
+public static IActionResult Run(HttpRequest req, TraceWriter log)
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Resources/Functions/WarmUp/run.csx]: Are these unrelated changes?

File: src/WebJobs.Script.WebHost/web.InProcess.xdt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
+  <system.webServer>
+    <handlers>
+      <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" xdt:Transform="Replace" xdt:Locator="Match(name)" />
+    </handlers>
+    <aspNetCore processPath=".\Microsoft.Azure.WebJobs.Script.WebHost.exe" arguments="" stdoutLogEnabled="true" stdoutLogFile="d:\home\logfiles\stdout" hostingModel="inprocess" xdt:Transform="Replace" xdt:Locator="Match(processPath)" />
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/web.InProcess.xdt]: We had explicitly disabled stdout logging in the past, do we need to have it enabled? This will cause files to be written to the slow file system.

File: build.ps1 (hunk 1 above)
Comment (safihamid) [reply build.ps1]: Yes, with CrossGen, we will need to recreate pdbs for .ni.dlls

File: build.ps1 (hunk 8 above)
Comment (safihamid) [reply build.ps1]: yes hence my comment about win-x64 or else 64bit V2 functions will be broken

File: src/WebJobs.Script.WebHost/Resources/Functions/WarmUp/run.csx (hunk 10 above)
Comment (safihamid) [reply src/WebJobs.Script.WebHost/Resources/Functions/WarmUp/run.csx]: We wanted these so the warmup/placeholder actually loads some more relevant dotnet core assemblies from disk and ngen to help with cold startup.

File: build.ps1 (hunk 3 above)
Comment (alrod) [reply build.ps1]: This PR do not address x64, this will be the next step.

File: src/WebJobs.Script.WebHost/Program.cs (hunk 9 above)
Comment (alrod) [reply src/WebJobs.Script.WebHost/Program.cs]: As discussed we leave it as is.

Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Program.cs]: 👍 for others, verified and this is a noop when not in proc

File: build.ps1 (hunk 3 above)
Comment (alrod) [reply build.ps1]: It's to hard to do. Leave this as is as discussed.

File: build.ps1 (hunk 8 above)
Comment (alrod) [reply build.ps1]: As now we are refer to 2.1 official release we can use "shared" model. Self-contained param is removed.

File: src/WebJobs.Script.WebHost/Resources/Functions/WarmUp/run.csx (hunk 10 above)
Comment (alrod) [reply src/WebJobs.Script.WebHost/Resources/Functions/WarmUp/run.csx]: This is just update to V2 new code for warm up function.
